pkg/apis/core/validation/validation_test.go

```diff
@@ -25002,6 +25003,120 @@ func TestValidateHostUsers(t *testing.T) {
+				HostnameOverride:  ptr.To("custom-host"),
+			},
+			expectedErrs: field.ErrorList{
+				field.Forbidden(field.NewPath("spec").Child("hostnameOverride"), "may not be specified when setHostnameAsFQDN is true"),
+			},
+		},
+		{
+			name: "SetHostnameAsFQDN=true and HostnameOverride is nil should not error",
+			spec: core.PodSpec{
+				SetHostnameAsFQDN: &trueVal,
+				HostnameOverride:  nil,
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "SetHostnameAsFQDN=false and HostnameOverride is set should not error",
+			spec: core.PodSpec{
+				SetHostnameAsFQDN: &falseVal,
+				HostnameOverride:  ptr.To("custom-host"),
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "SetHostnameAsFQDN is nil and HostnameOverride is set should not error",
+			spec: core.PodSpec{
+				SetHostnameAsFQDN: nil,
+				HostnameOverride:  ptr.To("custom-host"),
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "HostNetwork=false and HostnameOverride is set should not error",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To("custom-host"),
+				SecurityContext: &core.PodSecurityContext{
+					HostNetwork: false,
+				},
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "HostNetwork is not set and HostnameOverride is set should not error",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To("custom-host"),
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "HostNetwork=true and HostnameOverride is set should error",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To("custom-host"),
+				SecurityContext: &core.PodSecurityContext{
+					HostNetwork: true,
+				},
+			},
+			expectedErrs: field.ErrorList{
+				field.Forbidden(field.NewPath("spec").Child("hostnameOverride"), "may not be specified when hostNetwork is true"),
+			},
+		},
+		{
+			name: "Set HostnameOverride, but not RFC 1123 DNS subdomain should error",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To("Not-RFC1123"),
+			},
+			expectedErrs: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("hostnameOverride"), "Not-RFC1123", ""),
+			},
+		},
+		{
+			name: "Set HostnameOverride, but longer than 64 characters",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To(strings.Repeat("a", 65)),
+			},
+			expectedErrs: field.ErrorList{
+				field.TooLong(field.NewPath("spec").Child("hostnameOverride"), "", 64),
+			},
+		},
+		{
+			name: "Set HostnameOverride, equal 64 characters",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To(strings.Repeat("a", 64)),
+			},
+			expectedErrs: field.ErrorList{},
+		},
+		{
+			name: "Set HostnameOverride, less than 64 characters",
+			spec: core.PodSpec{
+				HostnameOverride: ptr.To(strings.Repeat("a", 1)),
+			},
+			expectedErrs: field.ErrorList{},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			errs := validatePodHostName(&tt.spec, field.NewPath("spec"))
+			matcher := field.ErrorMatcher{}.ByType().ByField().ByOrigin()
```

Comment:
If you add `ByDetailSubstring()` here then you can differentiate Forbidden by the reason with a substring instead of the full string (less brittle).